prot.c

```diff
@@ -1,15 +1,28 @@
+#include "prot.h"
 #include <sys/types.h>
-#include <unistd.h>
 #include <grp.h>
-#include "prot.h"
+#include <unistd.h>
 
-int prot_gid(gid_t gid)
+#ifndef GROUP_COUNT
+/* use a sane default, but one can override it from outside if there are
+ * higher requirements. */
+#define GROUP_COUNT 8
+#endif
+int prot_gids(const char *user, gid_t gid)
 {
-  if (setgroups(1,&gid) == -1) return -1;
-  return setgid(gid); /* _should_ be redundant, but on some systems it isn't */
-}
+  gid_t gids[GROUP_COUNT];
```

Comment:
I guess this is a little space-wasteful in the `prot_gid()` case, but meh

Reply:
For a 32 bit gid_t these are 32 byte of stack space - costs us basically nothing. There are worse things that happen on the stack and it is freed upon exit of this function, so who cares?